localstack/services/stepfunctions/asl/component/state/state_execution/state_map/iteration/iterator/distributed_iterator_worker.py

```diff
@@ -0,0 +1,137 @@
+        self._map_run_record.item_counter.running.count()
+
+        self._map_run_record.execution_counter.total.count()
+        self._map_run_record.execution_counter.running.count()
+
+        job_output = None
+        try:
+            env.context_object_manager.context_object["Map"] = Map(
+                Item=Item(Index=job.job_index, Value=job.job_input)
+            )
+
+            env.inp = job.job_input
+            env.stack.append(env.inp)
+            self._eval_input(env_frame=env)
+
+            job.job_program.eval(env)
+
+            # TODO: verify behaviour with all of these scenarios.
+            end_program_state: ProgramState = env.program_state()
+            if isinstance(end_program_state, ProgramError):
+                self._map_run_record.execution_counter.failed.count()
+                self._map_run_record.item_counter.failed.count()
+                job_output = None
+            elif isinstance(end_program_state, ProgramStopped):
+                self._map_run_record.execution_counter.aborted.count()
+                self._map_run_record.item_counter.aborted.count()
+            else:
+                self._map_run_record.item_counter.succeeded.count()
+                self._map_run_record.item_counter.results_written.count()
+
+                self._map_run_record.execution_counter.succeeded.count()
+                self._map_run_record.execution_counter.results_written.count()
+                self._map_run_record.execution_counter.running.offset(-1)
+
+                job_output = env.inp
+
+        except EvalTimeoutError:
+            self._map_run_record.item_counter.timed_out.count()
+
+        except FailureEventException:
+            self._map_run_record.item_counter.failed.count()
+
+        except Exception:
+            self._map_run_record.item_counter.failed.count()
+
+        finally:
+            self._map_run_record.item_counter.running.offset(-1)
+            job.job_output = job_output
+
+    def _eval_pool(self, job: Optional[Job], worker_frame: Environment) -> None:
+        # Note: the frame has to be closed before the job, to ensure the owner environment is correctly updated
+        #  before the evaluation continues; map states await for job termination not workers termination.
+        if job is None:
+            self._env.delete_frame(worker_frame)
+            return
+
+        # Evaluate the job.
+        job_frame = worker_frame.open_frame()
+        self._eval_job(env=job_frame, job=job)
+        worker_frame.delete_frame(job_frame)
+
+        # Evaluation terminates here due to exception in job.
+        if isinstance(job.job_output, Exception):
+            self._env.delete_frame(worker_frame)
+            self._job_pool.close_job(job)
+            return
+
+        # Worker was stopped.
+        if self.stopped():
+            self._env.delete_frame(worker_frame)
+            self._job_pool.close_job(job)
+            return
```

Comment:
nit: could be combined with an or statement but I'd leave that up to you. The current state might be more explicit and easier to read in the future